### PR TITLE
Auto-detect OpenSSL 1.1

### PR DESCRIPTION
**Pull request by wilzbach**
This is ugly, because dub is quite limited in this regard.
However, the `preBuildsCommands` + `rdmd` has been successfully used for a couple of
repositories at dlang-community already.

Also, I don't think this will work on Windows, but if the detection fails, the current behavior (i.e. using OpenSSL 1.0) is maintained.
For Posix-systems this should help a people a lot though and allows a good first five minutes experience.

edit: enabled it only for Posix systems

### CODE REVIEW

File: tls/dub.sdl [hunk 1]
```diff
@@ -21,6 +21,15 @@ configuration "openssl-mscoff" {
 
 configuration "openssl" {
 	sourceFiles "../lib/win-i386/eay.lib" "../lib/win-i386/ssl.lib" platform="windows-x86-dmd"
+	sourceFiles "openssl_version.d"
+	preBuildCommands `rdmd --eval='
```
[the remark this answers is not in the document]
Comment (wilzbach) [reply tls/dub.sdl]: Yes. In fact it will even prefer ldc if ldc is in the same directory.

[the remark this answers is not in the document]
Comment (Geod24) [reply tls/dub.sdl]: If this script fails, will it fail the build ?

[the remark this answers is not in the document]
Comment (wilzbach) [reply tls/dub.sdl]: Yes, but that's why the idea is to softly fail to `0.0.0` if no version could be detected. I also added a try/catch to be sure and enabled it for Posix only.

File: .travis.yml [hunk 2]
```diff
@@ -24,19 +24,19 @@ d:
...
   - ldc-1.7.0
   - ldc-1.6.0
   - ldc-1.5.0
   - ldc-1.4.0
+  - dmd-2.080.1
   - dmd-2.079.1
   - dmd-2.078.3
   - dmd-2.077.1
   - dmd-2.076.1
-  - dmd-2.075.1
```
Comment (wilzbach) [inline .travis.yml]: `rdmd --eval` doesn't work with 2.075: https://travis-ci.org/vibe-d/vibe.d/jobs/405854762

**Comment by WebFreak001** [comment]
the new mixin configs based on command output thingy that's being (been?) merged into dub could be useful here right?

**Comment by jacob-carlborg** [comment]
> the new mixin configs based on command output thingy that's being (been?) merged into dub could be useful here right?

Do you have any link for that?

**Comment by WebFreak001** [comment]
@jacob-carlborg https://github.com/dlang/dub/pull/1453

**Comment by bubnenkoff** [comment]
Yes, it's working on Windows 10 fine. At last a got code compile.

**Comment by wilzbach** [comment]
> the new mixin configs based on command output thingy that's being (been?) merged into dub could be useful here right?

Yep, but it's uncertain whether this will be approved (and it will take quite a while until this dub version is available/used everywhere). In the meantime, this trick works and removes the TLS configuration pain for Posix.

**Comment by wilzbach** [comment]
@s-ludwig: what's your opinion on this? Or can we finally move ahead with this?
It's been quite a while...

**Comment by s-ludwig** [comment]
We should definitely move forward, I wanted to do that for months now actually...

I mainly have two points:
- Splitting up the "openssl" configuration is a potentially breaking change, what was the reason for that instead of sticking `platform="..."` on each declaration as required?
- I'd probably go ahead one step further and make 1.1 the default if detection fails (adding an "openssl-1.0" configuration to force 1.0)

**Comment by wilzbach** [comment]
> Splitting up the "openssl" configuration is a potentially breaking change, what was the reason for that instead of sticking platform="..." on each declaration as required?

This:

```
Warning: Multiple configurations with the name "openssl" are defined in package "vibe-d:tls". This will most likely cause configuration resolution issues.
```

> I'd probably go ahead one step further and make 1.1 the default if detection fails (adding an "openssl-1.0" configuration to force 1.0)

Awesome! Done.

**Comment by s-ludwig** [comment]
What I mean is this:
```
configuration "openssl" {
	dependency "openssl" version="~>1.0"

	// Windows
	sourceFiles "../lib/win-i386/eay.lib" "../lib/win-i386/ssl.lib" platform="windows-x86-dmd"

	// Posix
	sourceFiles "openssl_version.d" platform="posix"
	preBuildCommands `rdmd --eval='
	auto dir = environment.get("DUB_PACKAGE_DIR");
	if (dir.buildPath("tls").exists)  {
		dir = dir.buildPath("tls");
	}
	auto opensslVersion = "0.0.0";
	try {
		const res = execute(["openssl", "version"]).output;
		if (res.canFind("OpenSSL ")) {
			opensslVersion = res.splitter(" ").dropOne.front.filter!(not!(std.uni.isAlpha)).text;
		}
	} catch (Exception e) { writeln("Warning: ", e); }
	text("module openssl_version;\nenum OPENSSL_VERSION=\"", opensslVersion, "\";").
		toFile(dir.buildPath("openssl_version.d"));
	'` platform="posix"
}
```

I think this should work fine.

Looks like you accidentally added some `.orig` files in the last commit.

**Comment by wilzbach** [comment]
> I think this should work fine.

Ah I see. Thanks! Updated.

> Looks like you accidentally added some .orig files in the last commit.

Yeah I was in a bit of a hurry. Sorry. Fixed.

**Comment by s-ludwig** [comment]
Upgraded and fixed the Windows side to work with OpenSSL 1.1, too. Let's hope it passes now.

**Comment by wilzbach** [comment]
Awesome. Thanks a lot! I guess we can release 0.8.5 fairly soon then?

**Comment by s-ludwig** [comment]
Definitiely! I'd just try wait for #2247 before tagging the beta (but speaking of which, I'll just tag a new alpha right now).

**Comment by bubnenkoff** [comment]
Can't build latest vibed on Windows 10 with follow error:
```
`` C:\Users\bubnenkov\AppData\Local\dub\packages\vibe-d-0.8.5-alpha.3\vibe-d\tls\vibe\stream\openssl.d(55,8): Error: variable `vibe.stream.openssl.OPENSSL_VERSION` conflicts with variable vibe.stream.openssl.OPENSSL_VERSION at C:\Users\bubnenkov\AppData\Local\dub\packages\vibe-d-0.8.5-alpha.3\vibe-d\tls\vibe\stream\openssl.d(41,7) ``
```

**Pull request merged**